### PR TITLE
Policy management

### PR DESCRIPTION
**Pull request by marcos-albornoz**
This PR contains all the changes related to the support for Policy Management

### CODE REVIEW

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import com.venafi.vcert.sdk.policy.domain.PolicySpecification;
+
+import java.io.File;
+import java.net.URL;
+
+/**
+ * The following test is to show how to use the policy management feature
+ * in order to create/update and get a policy for both TPP and Cloud (OutagePredict)
+ */
+public class PolicyManagementJsonTest {
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java:21]: Class name different from file name?

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            //and http://tutorials.jenkov.com/java-json/jackson-objectmapper.html
+
+            ObjectMapper mapper = new ObjectMapper();
+            mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+            mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+            mapper.enable(SerializationFeature.INDENT_OUTPUT);
+
+            PolicySpecification policySpecification = mapper.readValue( new File(json_source_file), PolicySpecification.class);
+
+            //2. Get a VCertClient. For this time, it will used a VCertClient for Cloud.
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java:44]: Grammar:
> For this time, it is going to use a VCertClient for Cloud.

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+            Config config = Config.builder()
+                    .connectorType(ConnectorType.CLOUD)
+                    .build();
+
+            VCertClient client = new VCertClient(config);
+            client.authenticate(auth);
+
+            //3. Use the VCertClient method setPolicy() to set a Policy.
+            // If the the policy doesn't exist then that will be created.
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java:57]: Same issue with previous example

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            // If the the policy exists then it will updated.
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java:58]: Same issue with previous example

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import com.venafi.vcert.sdk.endpoint.ConnectorType;
+import com.venafi.vcert.sdk.policy.domain.PolicySpecification;
+
+import java.io.File;
+
+/**
+ * The following test is to show how to use the policy management feature
+ * in order to create/update and get a policy for both TPP and Cloud (OutagePredict)
+ */
+public class PolicyManagementYamlTest {
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java:20]: Same issue with class name

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            // You can learn more about Jackson parser in https://github.com/FasterXML/jackson
+            //and http://tutorials.jenkov.com/java-json/jackson-objectmapper.html
+
+            ObjectMapper mapper = new ObjectMapper(new YAMLFactory());
+            mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+            mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+
+            PolicySpecification policySpecification = mapper.readValue( new File(yaml_source_file), PolicySpecification.class);
+
+            //2. Get a VCertClient. For this time, it will used a VCertClient for Cloud.
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java:42]: Same issue as previous examples

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+            Config config = Config.builder()
+                    .connectorType(ConnectorType.CLOUD)
+                    .build();
+
+            VCertClient client = new VCertClient(config);
+            client.authenticate(auth);
+
+            //3. Use the VCertClient method setPolicy() to set a Policy.
+            // If the the policy doesn't exist then that will be created.
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java:55]: Same issue as previous examples

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            // If the the policy exists then it will updated.
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java:56]: Same issue as previous examples

File: examples/com/venafi/vcert/sdk/example/policymanagement/policy_specification.json [hunk 9]
```diff
@@ -0,0 +1,68 @@
+{
+  "name": "",
+  "policy": {
+    "domains": [
+      "venafi.com",
+      "kwan.com"
+    ],
+    "maxValidDays": 120,
+    "wildcardAllowed": false,
+    "certificateAuthority": "ENTRUST\\Entrust\\STANDARD_SSL",
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/policy_specification.json:10]: Do this CA exists? If the file is expected to be used, then set a valid CA.

File: examples/com/venafi/vcert/sdk/example/policymanagement/policy_specification.yaml [hunk 10]
```diff
@@ -0,0 +1,50 @@
+---
+name: ''
+policy:
+  domains:
+    - venafi.com
+    - kwan.com
+    - yaml.com
+  wildcardAllowed: true
+  maxValidDays: 120
+  certificateAuthority: ENTRUST\\Entrust\\STANDARD_SSL
```
Comment (rvelaVenafi) [inline examples/com/venafi/vcert/sdk/example/policymanagement/policy_specification.yaml:10]: Same doubt. Does the CA exists?

File: src/main/java/com/venafi/vcert/sdk/endpoint/Authentication.java [hunk 11]
```diff
@@ -48,7 +48,7 @@ public Authentication(String user, String password, String accessToken, String r
 	@Override
 	public String toString() {
 		return Authentication.class.getSimpleName() + "(user=" + user + ", apiKey=" + apiKey
-				+ ", password=" + (!password.isEmpty() ? "****" : "not set") + ")";
+				+ ", password=" + (password!=null && !password.isEmpty() ? "****" : "not set") + ")";
 	}
 
 }
```
Comment (rvelaVenafi) [inline src/main/java/com/venafi/vcert/sdk/endpoint/Authentication.java:54]: No new line at end of file.

File: src/main/java/com/venafi/vcert/sdk/policy/converter/cloud/CloudPolicyToPolicyConverter.java [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+        if( recommendedSettings.key() != null ) {
+            if( recommendedSettings.key().type() != null )
+                getDefaultsKeyPairFromPolicySpecification(policySpecification).keyType(recommendedSettings.key().type());
+
+            if( recommendedSettings.key().length() != null )
+                getDefaultsKeyPairFromPolicySpecification(policySpecification).rsaKeySize(recommendedSettings.key().length());
+        }
+    }
+
+}
```
Comment (rvelaVenafi) [inline src/main/java/com/venafi/vcert/sdk/policy/converter/cloud/CloudPolicyToPolicyConverter.java:186]: No new line at end of file

File: src/main/java/com/venafi/vcert/sdk/policy/converter/tpp/TPPPolicySpecificationValidator.java [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+            if(policyKeyPair != null
+                    && policyKeyPair.serviceGenerated() != null && defaultsKeyPair.serviceGenerated() != null
+                    && !policyKeyPair.serviceGenerated().equals(defaultsKeyPair.serviceGenerated())) {
+                    throw new VCertException(String.format(DEFAULT_ATTRIBUTE_DOESNT_MATCH_EXCEPTION_MESSAGE, PolicySpecificationConst.ATT_DEFAULTS_KEYPAIR_SERVICE_GENERATED_TYPE, PolicySpecificationConst.ATT_POLICY_KEYPAIR_SERVICE_GENERATED_TYPE));
+            }
+        }
+    }
+
+}
```
Comment (rvelaVenafi) [inline src/main/java/com/venafi/vcert/sdk/policy/converter/tpp/TPPPolicySpecificationValidator.java:183]: No new line at end of file

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+import com.venafi.vcert.sdk.endpoint.Authentication;
+import com.venafi.vcert.sdk.endpoint.ConnectorType;
+import com.venafi.vcert.sdk.policy.domain.*;
+import com.venafi.vcert.sdk.policy.domain.Policy;
+
+/**
+ * The following test is to show how to use the policy management feature
+ * in order to create/update and get a policy for both TPP and Cloud (OutagePredict)
+ */
+public class PolicyManagementBuilderTest {
```
[the remark this answers is not in the document]
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java:19]: mmmm, I see that the refactoring name IntelliJ action only refactored the name of the file, not the name of the class.

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java (hunk 1 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java:21]: mmmm, I see that the refactoring name IntelliJ action only refactored the name of the file, not the name of the class.

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java (hunk 5 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java:20]: mmmm, I see that the refactoring name IntelliJ action only refactored the name of the file, not the name of the class.

File: examples/com/venafi/vcert/sdk/example/policymanagement/policy_specification.json (hunk 9 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/policy_specification.json:10]: Yes, it exists in Cloud

File: examples/com/venafi/vcert/sdk/example/policymanagement/policy_specification.yaml (hunk 10 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/policy_specification.yaml:10]: Yes, it exists in Cloud

File: src/main/java/com/venafi/vcert/sdk/endpoint/Authentication.java (hunk 11 above)
Comment (marcos-albornoz) [reply src/main/java/com/venafi/vcert/sdk/endpoint/Authentication.java:54]: In fact that doesn't affect the functionallity of the class

File: src/main/java/com/venafi/vcert/sdk/policy/converter/cloud/CloudPolicySpecificationValidator.java [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+                if( !CloudKeyPairEnums.containsRsaKeySize( defaultRsaKeySize ))
+                    throw new VCertException(String.format(DEFAULT_ATTRIBUTE_DOESNT_MATCH_WITH_ACCEPTED_VALUES_EXCEPTION_MESSAGE, PolicySpecificationConst.ATT_DEFAULTS_KEYPAIR_RSA_KEY_SIZE));
+
+                if(policyKeyPair != null && !Arrays.stream(policyKeyPair.rsaKeySizes()).anyMatch(defaultRsaKeySize::equals))
+                    throw new VCertException(String.format(DEFAULT_ATTRIBUTE_DOESNT_MATCH_EXCEPTION_MESSAGE, PolicySpecificationConst.ATT_DEFAULTS_KEYPAIR_RSA_KEY_SIZE, PolicySpecificationConst.ATT_POLICY_KEYPAIR_RSA_KEY_SIZES));
+            }
+        }
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (marcos-albornoz) [reply src/main/java/com/venafi/vcert/sdk/policy/converter/cloud/CloudPolicySpecificationValidator.java:189]: In fact that doesn't affect the functionallity of the class

File: src/main/java/com/venafi/vcert/sdk/policy/converter/cloud/CloudPolicyToPolicyConverter.java (hunk 12 above)
Comment (marcos-albornoz) [reply src/main/java/com/venafi/vcert/sdk/policy/converter/cloud/CloudPolicyToPolicyConverter.java:186]: In fact that doesn't affect the functionallity of the class

File: src/main/java/com/venafi/vcert/sdk/policy/converter/tpp/TPPPolicySpecificationValidator.java (hunk 13 above)
Comment (marcos-albornoz) [reply src/main/java/com/venafi/vcert/sdk/policy/converter/tpp/TPPPolicySpecificationValidator.java:183]: In fact that doesn't affect the functionallity of the class

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+                                    .reuseAllowed(true)
+                                    .build())
+                            .subjectAltNames( SubjectAltNames.builder()
+                                    .dnsAllowed(false)
+                                    .emailAllowed(true)
+                                    .build())
+                            .build())
+                    .build();
+
+            //2. Get a VCertClient. For this time, it will used a VCertClient for TPP.
```
[the remark this answers is not in the document]
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+            Config config = Config.builder()
+                    .connectorType(ConnectorType.TPP_TOKEN)
+                    .baseUrl(baseUri)
+                    .build();
+
+            VCertTknClient client = new VCertTknClient(config);
+            client.getAccessToken(auth);
+
+            //3. Use the VCertClient method setPolicy() to set a Policy.
+            // If the the policy doesn't exist then that will be created.
```
[the remark this answers is not in the document]
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+            // If the the policy exists then it will updated.
```
[the remark this answers is not in the document]
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+            client.setPolicy(policyName, policySpecification);
+
+            //4. You can get the Policy which you created/updated using the getPolicy method
+            PolicySpecification policyTemp = client.getPolicy(policyName);
+
+            //5. then use it to write it in Yaml format.
```
[the remark this answers is not in the document]
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementBuilderExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java (hunk 1 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java (hunk 2 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java (hunk 4 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java (hunk 3 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementJsonExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java (hunk 5 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java (hunk 6 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java (hunk 7 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java]: it's done

File: examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java (hunk 8 above)
Comment (marcos-albornoz) [reply examples/com/venafi/vcert/sdk/example/policymanagement/PolicyManagementYamlExample.java]: it's done

**Review by angelmoo — APPROVED** [review]

**Review by rvelaVenafi — APPROVED** [review]

**Pull request merged**